Add github actions workflow to test SRWA build

robgpita opened the pull request:
## DESCRIPTION OF CHANGES: 

Currently, there are no Continuous Integration (C.I.) tests for this repository.  C.I. tests are considered a software engineering best practice, with numerous benefits  to developers as well as the code-base. This PR adds the first step in implementing a C.I. pipeline (building the ufs-srweather-app).  Forthcoming work will add another GitHub Actions workflow file which will automate the running of Workflow End To End tests. 

## TESTS CONDUCTED: 

A build test accomplished via the GHA workflow file was done on an AWS Parallel cluster configured as a self-hosted runner with all of the necessary dependencies, including `hpc-stack` with versions matched as closely as possible to srw_common (different patch of g2).  As versions of dependent libraries change, the self-hosted runner will need to be updated, reflecting those changes. 

A [successful GitHub actions workflow run](https://github.com/robgonzalezpita/regional_workflow/actions/runs/2235626241) has been accomplished, on my fork. Once a GitHub Actions Self-Hosted runner is configured and assigned to the NOAA-GSL fork, the proposed GitHub Actions Workflow will run on it. 

## DEPENDENCIES:
This GitHub Action Workflow is run on a self-hosted runner, which must be assigned to this repository. A machine configured as a Github Actions self-hosted runner is necessary to run this workflow.  

[PR #152](https://github.com/NOAA-GSL/ufs-srweather-app/pull/152) for the ufs-srweather-app is a concurrent pull request. While not a dependency, once that one has been accepted, the workflow can be simplified in utilizing those changes. 

## DOCUMENTATION:

[PR #152](https://github.com/NOAA-GSL/ufs-srweather-app/pull/152) in the `ufs-srweather-app` includes a `docs/UsersGuide/source/ContinuousIntegration.rst` file.

## CONTRIBUTORS: 
@christinaholtNOAA @christopherwharrop-noaa @venitahagerty

Commit:
Add github actions workflow to run on self-hosted runner

christinaholtNOAA commented on .github/workflows/build_SRWA.yml:
```diff
@@ -0,0 +1,89 @@
+    # Auxillary repository needed to insert env files to run CI tests in the short term 
+    name: Clone RRFS CI repo
+    # This job will run on AWS Parallel Cluster head node
+    runs-on: self-hosted
```
@robgonzalezpita I assume the failure on this task means that we need to work together to get the self-hosted runner set up on this branch.

robgpita replied:
The failure is due to not having a self-hosted runner set up for this repository. Yes, we do need to work together on configuring a runner to pick up jobs initiated by this workflow.

christinaholtNOAA commented on .github/workflows/build_SRWA.yml:
```diff
@@ -0,0 +1,89 @@
+    runs-on: self-hosted
+    steps:
+    
+    - name: Checkout robgonzalezpita/rrfs-ci-pcluster Repo
```
We should probably work on transferring this to the NOAA-GSL organization.

christinaholtNOAA commented on .github/workflows/build_SRWA.yml:
```diff
@@ -0,0 +1,89 @@
+            echo 'branch exists!'
+            git checkout ${{ github.ref_name }}
+          else
+            git checkout rrfs_ci
```
Is there a way to keep from hard-coding this so deep in the workflow? Maybe even just pulling it out to a more prominent place will help us remember to change it to the most appropriate default branch name?

robgonzalezpita replied:
Potentially, I'll look into this a little bit more. I'm not quite sure where a more prominent place would be. Are you thinking more prominent just to remember to change rrfs_ci back to develop? Or are you thinking we'll want to have a variable for the branch we're submitting the PR against?

robgpita replied:
Yep, I changed `rrfs_ci` to `${{ github.base_ref }}`

christinaholtNOAA commented on .github/workflows/build_SRWA.yml:
```diff
@@ -0,0 +1,89 @@
+      # Specify branch used for testing, if there is not a similar named branch in ufs-SRWA
+      - name: Checkout SRWA 
+        run: |
+          cd ../
```
We will likely want a bit more information about the directory structure here. Consider using full paths, or a doc block to explain the topology that we'll be traversing here. Please address the relative paths in the next sections similarly.

robgpita replied:
The full path used will be dependent on the location we install the self hosted runner. I'm holding off on this one for now until we get the NOAA-GSL self-hosted runner configured.

robgpita replied:
I've added a couple of comments addressing the reason why relative paths are used and the topology we're traversing.

christinaholtNOAA commented on .github/workflows/build_SRWA.yml:
```diff
@@ -0,0 +1,89 @@
+       
+      - name: Insert this Version of Regional Workflow
+        run: |
+          cp -r ${{ github.workspace }} ../${{ github.sha }}/ufs-srweather-app/regional_workflow
```
How does regional_workflow get there? Does it have it's own path for every PR on the runner?

robgpita replied:
Every time this workflow is triggered, the runner checks out the repository in `github.workspace`. The way it is currently set up, yes, the repo has its own path every time the workflow is initiated. 

The reason I've set it up this way is to handle concurrent pull requests.  There very well could be a better way to account for multiple PRs.  I'm quite open to alternative approaches.

christinaholtNOAA commented on .github/workflows/build_SRWA.yml:
```diff
@@ -0,0 +1,89 @@
+        run: |
+          cd ../${{ github.sha }}/ufs-srweather-app/test/
+          ./build.sh >& build_test.out
+          cat build_test*
```
We can get this in real time with this command (although I'm not sure if it will work in a GHA):

```
./build.sh 2>&1 | tee build_test.out
```

robgpita replied:
I'll try using that line instead, and replace if it works in GHA.

christinaholtNOAA reviewed (changes requested):
@robgonzalezpita This is nice work. I left some comments/suggestions below.

Commit:
Checkout base branch of SRWA, modi

If corresponding branch is not found in SRWA repository, checkout the base branch.  Change how  ./build.sh is printed to stdout.

Commit:
Remove duplicate calling of build.sh

Commit:
Merge remote-tracking branch 'origin/rrfs_ci' into feature/CI_build_test

Commit:
Remove insertion of UFS SRWA env files from seperate repository, print github.refs for debugging on web interface

Commit:
Refactor build_SRWA.yml to clone concurrent PR from the same author and branch name in the SRWA, add final job to display build status in web interface

Commit:
Add Comment addressing relative file paths, add new line at the end of file

Commit:
Add more detailed comments

robgpita reviewed (commented):
I believe I've addressed all of the old requested changes. Please re-review and suggest any new changes.

christinaholtNOAA commented on .github/workflows/build_SRWA.yml:
```diff
@@ -0,0 +1,119 @@
+  # (Cloning repositories outside of the GITHUB_WORKSPACE directory is not supported with actions/checkout@v3, so it needs to be moved into the parent directory seperately)
+  # The SRWA is cloned within the GITHUB_WORKSPACE (default location of where 'run' commands are executed ), we need to move its location up one directory, as the regional_workflow is nested within the SRWA.
+  Clone_SRWA:
+    name: Clone SRWA and move into parent directory
```
Why does it _need_ to move up one directory? Can you not clone SRW where you want it, and then clone regional_workflow inside that clone using the `path` option for both?

robgpita replied:
@christinaholtNOAA, you pose some great questions, and this a much better approach. I updated the GHA workflow to clone SRW and place the regional_workflow inside of that clone using the `path` option.

Commit:
Try cloning RRFS using actions/checkout within the srwa clone

Commit:
Fix merge

Commit:
Remove outdated comments
